### PR TITLE
add waitstatus and returncode job attributes to JobID class and flux-jobs

### PR DESCRIPTION
**Pull request by grondo**
This PR adds a new `waitstatus` attribute to job-info, and makes it available via the `job-info.list` interface.  This value is simply the cached `status` from the job `finish` event.

The `waitstatus` is then made available as a property in the `JobInfo` class, along with a new computed `returncode` property. The `returncode` is the exit code equivalent of the `waitstatus` if the job shell exited (i.e. `os.WEXITSTATUS(waitstatus)`, or the negative signal number if the shell was terminated due to a signal (e.g. the same as Python's [subprocess.Popen.returncode](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.returncode)). If the job was cancelled before it ever started, then the special value `-128` is returned.

This scheme allows all inactive jobs to have a valid `returncode` which is non-zero if the job did not exit successfully.

### CODE REVIEW

File: t/t2800-jobs-cmd.t [hunk 1]
```diff
@@ -639,6 +639,26 @@ test_expect_success 'flux-jobs --format={status},{status_abbrev} works' '
...
+test_expect_success 'flux-jobs --format={waitstatus},{returncode}' '
+	FORMAT="{waitstatus:h},{returncode:h}" &&
+	flux jobs --filter=pending,running -no "$FORMAT" > returncodePR.out &&
+	flux jobs --filter=inactive -no "$FORMAT" > returncodeI.out &&
+	test_debug "echo active:; cat returncodePR.out" &&
+	test_debug "echo inactive:; cat returncodeI.out" &&
+	countPR=$(grep -c "^-,-$" returncodePR.out) &&
+	test_debug "echo active got $countPR, want $(state_count sched run)" &&
+	test $countPR -eq $(state_count sched run) &&
+	count=$(grep -c "^32512,127$" returncodeI.out) &&
```
Comment (chu11) [inline t/t2800-jobs-cmd.t:651]: nit: should use different filename than `returncodeI.out` for each case

**Review by chu11 — COMMENTED** [review]
overall LGTM, but thinking we could use some `waitstatus` tests in `t2230-job-info-list.t`.

Also, in `t2230-job-inf-list` and `t2800-jobs-cmd`, perhaps we want to add a failed job that died to a signal, to add additional coverage.

**Comment by grondo** [comment]
> Also, in t2230-job-inf-list and t2800-jobs-cmd, perhaps we want to add a failed job that died to a signal, to add additional coverage.

I think we should leave this out until #2773 is solved.

File: t/t2800-jobs-cmd.t (hunk 1 above)
Comment (grondo) [reply t/t2800-jobs-cmd.t:651]: I don't follow. Are you suggesting separate test for `{returncode}` vs `{waitstatus}`, or separating failed completed and cancelled job listing into separate output files? (That would seem to unnecessarily slow down this test, which is already one of the longest in the testsuite)

Or am I missing something obvious?

**Comment by grondo** [comment]
> overall LGTM, but thinking we could use some waitstatus tests in t2230-job-info-list.t.

Do you mean just test that `flux job list` includes the `waitstatus` field? O/w most of the tests would be redundant I think...

File: t/t2800-jobs-cmd.t (hunk 1 above)
Comment (chu11) [reply t/t2800-jobs-cmd.t:651]: oh sorry, my bad ... thought you were outputting to `returncodeI.out` multiple times, but you are actually grepping the same output in `returncodeI.out` multiple times.

So everythings fine :-)

**Comment by chu11** [comment]
> Do you mean just test that flux job list includes the waitstatus field? O/w most of the tests would be redundant I think...

Yeah, just that `flux job list` includes a `waitstatus` field test.

I've been adding tests to both files for a long time, somewhat out of habit/routine, but they are admittedly redundant now now that `flux-jobs` has overtaken `flux job list`.  Perhaps we should remove a bunch of tests from `t2230-job-info-list` instead.  i can do that as a simple follow up PR.

**Comment by grondo** [comment]
> I think we should leave this out until #2773 is solved.

Actually I can use `flux jobs --from-stdin` to test this case, esp since we use `flux-jobs` to test the `JobInfo` class. I'll add that and the `waitstatus` check for flux-job list.

**Review by chu11 — APPROVED** [review]
given above discussion, no problems with this PR

**Comment by grondo** [comment]
Thanks, @chu11! I added a `flux jobs --from-stdin` test to ensure that the `returncode` is the negative signal number when `waitstatus` indicates the shell was terminated by a signal. I'll set MWP now.

**Pull request merged**